utils/s3/client.hh

```diff
@@ -30,10 +31,43 @@ namespace s3 {
 
 using s3_clock = std::chrono::steady_clock;
 
-struct range {
-    uint64_t off;
-    size_t len;
+class range {
+    friend struct fmt::formatter<range>;
+
+public:
+    range() noexcept = delete;
+    constexpr range(uint64_t offset, uint64_t len) noexcept : _offset(offset), _length(len) {
+        uint64_t t;
+        SCYLLA_ASSERT(!__builtin_add_overflow(_offset, len, &t));
+        SCYLLA_ASSERT(_offset + len <= _max_object_size);
+    }
+    explicit constexpr range(uint64_t offset) noexcept : _offset(offset), _length(std::nullopt) {}
+    constexpr range& operator+=(uint64_t offset) noexcept {
+        uint64_t t;
+        SCYLLA_ASSERT(!__builtin_add_overflow(_offset, offset, &t));
+        SCYLLA_ASSERT(_offset + offset <= _max_object_size);
+
+        _offset += offset;
+        if (_length) {
+            SCYLLA_ASSERT(!__builtin_sub_overflow(_length.value(), offset, &t));
+            _length.value() -= offset;
+        }
+        return *this;
+    }
+    bool operator==(const range& other) const noexcept { return std::tie(_offset, _length) == std::tie(other._offset, other._length); }
```

Comment:
Would default-generated <=> work as well?

Reply:
Why having operator== when the <=> is around?
And also the question about `operator<=> ... = default` still stands

Reply:
done